### ISSUE TITLE
DeepSlice web gives much better results than the offline version

### ISSUE THREAD

**Issue by paulbrodersen** [PolarBean/DeepSlice#43]
Hi, thanks for making DeepSlice and making it available for the wider community. When I initially stumbled across DeepSlice, I tried out the web version and was quite impressed with the registration results. I have since tried to replicate the results by running DeepSlice offline/locally but with no success. The model runs with no issues, the output XML/JSON file structure seems fine and imports fine into Quicknii, but the registration is clearly garbage. 

For example: 

Autofluorescence image of slice 1 (of 11)

![brain_slice_s001](https://github.com/PolarBean/DeepSlice/assets/8046146/493122b8-dbb5-4c06-89ed-12bed72d8634)

Offline result:

![brain_slice_s001-STPt_avg 2015](https://github.com/PolarBean/DeepSlice/assets/8046146/365b1c4d-3dad-43fb-9d7f-8e223670a1ce)

Online result (for comparison):

![brain_slice_s001-STPt_avg 2015](https://github.com/PolarBean/DeepSlice/assets/8046146/02ed82bf-c162-42d0-8a1c-b1f51ff9979e)

Any pointers towards potential issues or what to try next would be much appreciated. 

---

### Script used to create the offline results

```python
from DeepSlice import DSModel

folder = "/path/to/images/"
Model = DSModel("mouse")
Model.predict(folder, ensemble=True, section_numbers=True)
Model.enforce_index_spacing(section_thickness=150)
Model.propagate_angles()
Model.save_predictions(folder + 'deepslice_results')
```

### Settings used for getting the online results (=all boxes ticked)

![Screenshot from 2023-09-20 13-04-53](https://github.com/PolarBean/DeepSlice/assets/8046146/d349a4e0-a0a6-4486-ab6f-d159da9de6f7)

## Additional files 

[offline results XML](https://drive.google.com/file/d/1NuawWqG4eNq6iCklbtaxjiePYlRr8R9I/view?usp=sharing)

[online results XML](https://drive.google.com/file/d/1uIiKOv61JKJvxOoYmzQFZkrs7_i858y3/view?usp=sharing)

[original images](https://drive.google.com/drive/folders/1LPTV2XpsB0UJn9fUzDERk86OyFrCnzyv?usp=sharing)

**Comment by PolarBean** [comment]
Hey thanks for using DeepSlice! This is a great point that I need to add to the user documentation. section_thickness expects not only the thickness in microns but also the direction of cutting. with caudal to rostral being positive and rostral caudal being negative. In this case your sections are numbered rostral to caudal so then we simply specify 
```python
Model.enforce_index_spacing(section_thickness=-150)
```
and it will work as expected! 

Let me know how you go. (thank you for providing the images and json, it makes figuring out the issue very easy :) )

**Comment by paulbrodersen** [comment]
Dude. Duuuuuuude. I have been banging my head against the wall because of that missing minus sign all day yesterday. 
(In other words: thanks for the quick reply & the offline results are now just as good as the online results). 

Suggestion: make the orientation explicit in the function definition by using a second argument. E.g.:

```python
Model.enforce_index_spacing(section_thickness=150, indexing_from_caudal_to_rostral=False)
```

And internally:

```python
def enforce_index_spacing(self, section_thickness, indexing_from_caudal_to_rostral=True):
    ...
    If not indexing_from_caudal_to_rostral:
        section_thickness = -np.abs(section_thickness)
    ...
```

This is backwards compatible and much more explicit than having a sentence somewhere in the documentation that explains that a thickness is a signed number.

**Comment by PolarBean** [comment]
Yeah this is definitely an oversight on my behalf. If you make this change I will merge it, otherwise I’ll get on it soon :) I agree with it completely

**Comment by paulbrodersen** [comment]
After having read the code, I think my suggestion is maybe the quickest but potentially not the best way to make that change. 

As I understand it, you already determine the rostro-caudal orientation of the indices (a.k.a `direction`) in `enforce_index_order` -> `spacing_and_indexing.enforce_section_ordering`.  

1. An easy win (IMO) would be to expose the `direction` variable to the user: 

main.py: 

```python 
    def enforce_index_order(self, direction:str = None):
        """
        reorders the section depths (oy) in the predictions such that they align with the section indexes

        :param direction: The orientation in the rostro-caudal axis. One of 'caudal-rostro', 'rostro-caudal', or None.
                          Defaults to None, in which case the orientation is automatically determined.
        :type direction: str or None
        """
        self.predictions = spacing_and_indexing.enforce_section_ordering(
            self.predictions, direction
        )
```

spacing_and_indexing.py:

```python
def enforce_section_ordering(predictions, direction:str = None):
    """
    Ensures that the predictions are ordered by section number

    :param predictions: dataframe of predictions
    :type predictions: pandas.DataFrame
    :param direction: The orientation in the rostro-caudal axis. One of 'caudal-rostro', 'rostro-caudal', or None.
                      Defaults to None, in which case the orientation is automatically determined.
    :type direction: str or None
    :return: the input dataframe ordered by section number
    :rtype: pandas.DataFrame
    """
    predictions = predictions.sort_values(by=["nr"], ascending=True).reset_index(drop=True)
    if len(predictions) == 1:
        raise ValueError("Only one section found, cannot space according to index")
    if "nr" not in predictions:
        raise ValueError(
            "No section indexes found, cannot enforce index order. You likely did not run predict() with section_numbers=True"
        )
    else:
        predictions = predictions.reset_index(drop=True)
        depths = calculate_brain_center_depths(predictions)
        depths = np.array(depths)
        if direction is None:
            direction = determine_direction_of_indexing(depths)
        predictions["depths"] = depths

        temp = predictions.copy()
        if direction == "caudal-rostro":
            ascending = False
        if direction == "rostro-caudal":
            ascending = True
        if "bad_section" in temp:
            temp_good = temp[temp["bad_section"] == False].copy().reset_index(drop=True)
            temp_good_copy = temp_good.copy()
            temp_good_copy = temp_good_copy.sort_values(by=["depths"], ascending=ascending).reset_index(
                drop=True
            )
            temp_good["oy"] = temp_good_copy["oy"]

            predictions.loc[predictions["bad_section"] == False, "oy"] = temp_good["oy"].values
        else:
            temp = temp.sort_values(by=["depths"], ascending=ascending).reset_index(drop=True)


            predictions["oy"] = temp["oy"].values
    return predictions
```

2. Independent of whether you expose the `direction` variable to the user, this is the place in the code that should determine the sign of the `section_thickness` variable. The way that the code is currently structured makes that difficult, which to me suggests that `enforce_section_ordering` and `enforce_section_thickness` should maybe be combined. Alternatively, you could split the process into three parts, `set_rostro_caudal_orientation` setting a `Model.direction` attribute, and `enforce_section_ordering` and `enforce_section_thickness` then using this attribute in their computations.

**Comment by PolarBean** [comment]
oh thats a great idea. 
So we add a direction argument to the `enforce_index_spacing` method of `DSModel` and then passing this argument to `space_according_to_index`  and then finally to `enforce_section_ordering` 

I would advise against adding a Model.direction attribute since this would suggest the parameter is remembered for future registrations within the same session which might be confusing. 

The `direction` variable is set somewhat independently of  `section_thickness`  so we would still need to do something similar to your first suggestion. 

`enforce_section_ordering` is called by `space_according_to_index` and im not convinced they should be combined since they are quite distinct and independantly relatively atomic functions.  I think perhaps in the future I would like to seperate them further and not call `enforce_section_ordering`  from `space_according_to_index`  but rather call them sequentially in  `enforce_index_spacing` 

so in summary I agree mostly, we 
1. Expose the direction variable to the user and pass it through `enforce_index_spacing` -> `space_according_to_index` -> `enforce_section_ordering` 
2. Use the direction variable to set the sign of `section_thickness`  and do this in `space_according_to_index`

**Comment by paulbrodersen** [comment]
> im not convinced they should be combined since they are quite distinct and independantly relatively atomic functions

Agreed, as I didn't like my suggestion either. But you are clearly getting my broader point that this part of the code needs a little bit re-structuring. 

Otherwise, it's good code. You should advertise more, so more people hear about it and use it.

**Comment by PolarBean** [comment]
At the moment we are relying on word of mouth mainly!
